harp/reader.py

```diff
@@ -44,13 +45,37 @@ def __init__(
         self.read = read
 
 
+class RegisterMap(UserDict[str, RegisterReader]):
+    _address_map: dict[int, str]
+
+    def __init__(self, registers: dict[str, RegisterReader]) -> None:
+        super().__init__(registers)
+        self._address_map = {self[key].register.address: key for key in self.keys()}
+
+    def __getitem__(self, __key: Union[str, int]) -> RegisterReader:
+        if isinstance(__key, int):
+            super().__getitem__(self._address_map[__key])
+        elif isinstance(__key, str):
+            return super().__getitem__(__key)
+        else:
+            raise TypeError(f"key must be int or str, not {type(__key)}")
+
+    def __getattr__(self, __name: str) -> RegisterReader:
+        return self[__name]
+
```

Comment:
I think we should avoid defining `__getattr__` here so that the `RegisterMap` class works only as a pure dict. The code completion attributes were meant to be provided only at the level of the `DeviceReader` class.

```suggestion
```